File: LCOF/18.shan-chu-lian-biao-de-jie-dian-lcof/shan-chu-lian-biao-de-jie-dian-lcof.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteNode(head *ListNode, val int) *ListNode {
    dummy := &ListNode{Val: 0, Next: head} // 假头
    first := dummy // 双指针1
    second := dummy.Next // 双指针2
    for {
        if second.Val == val {
            first.Next = second.Next  //删除指定节点
            break
        }
        first = first.Next
        second = second.Next
    }

    return dummy.Next
}

File: LCOF/18.shan-chu-lian-biao-de-jie-dian-lcof/shan-chu-lian-biao-de-jie-dian-lcof.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteNode(ListNode* head, int val) {
        struct ListNode** indirect = &head; 
    //struct ListNode* temp;
    for(; *indirect; indirect = &((*indirect)->next)){
        if((*indirect)->val == val){
            //temp = *indirect;
            *indirect = (*indirect)->next;
            //free(temp);
            break;
        }
    }
    return head;
    }
};


/*
Linus Solution
http://wordaligned.org/articles/two-star-programming 
*/
